kernel/mem_domain.c

```diff
@@ -42,6 +43,13 @@ static bool sane_partition(const struct k_mem_partition *part,
 		if (last < parts[i].start || cur_last < part->start) {
 			continue;
 		}
+#if defined(CONFIG_MPU_REQUIRES_NON_OVERLAPPING_REGIONS)
+		/* Partitions overlap */
+		__ASSERT(false, "overlapping partitions <%x...%x>, <%x...%x>",
+			part->start, last,
+			parts[i].start, cur_last);
+		return false;
```

Comment:
Same pattern.  Log an error, don't assert for a runtime condition you expect to be handled.

Reply:
Oh wait. kernel/mem_domain.c does not use the Logger yet, so I can not leave a LOG_ERR message. Although I totally agree with no having ASSERTS when we do return error signals, I prefer to stick with the rule in this module (and in the very same functio actually) and leave the ASSERT.

We can definitely fill-in an issue to make mem_domain use Logger and also remove ASSERTS entirely when signal errors are propagated.
I will track that externally.

Reply:
Then change this to an _ASSERT(!IS_ENABLED(CONFIG_MPU_REQUIRES_NON_OVERLAPPING_REGIONS), "") maybe?  That would better express this as a configuration error and not a runtime one.

Or just printk, honestly.  No shame in printk.

Reply:
I concur with Andy

Reply:
Am I correct that the existing code only checks for consecutive overlapping memory partitions?

Reply:
No, @aurel32 it checks whether the current partition overlaps with any other partition in this domain.

Reply:
Hmm correct, sorry about the confusion.